### PR TITLE
[WIP]Improve scalability

### PR DESCRIPTION
**Pull request by Gnimuc**
Here some benchmark results:
```julia
using PkgBenchmark
julia> judge("Hungarian", "speed-up", "master")
INFO: Resolving dependencies for benchmark
INFO: No packages to install, update or remove
INFO: Running benchmarks...
Using benchmark tuning data in /Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/.tune.jld
INFO: No packages to install, update or remove
File results of this run? (commit=354ffe, resultsdir=/Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results) (Y/n) y
INFO: Results of the benchmark were written to /Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results/354ffe3f33ef4a0e1683d4aea6ad64485c8d6c53.jld
INFO: Reading results for 04d0fe from /Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "square matrix" => 8-element BenchmarkGroup([])

julia> showall(ans)
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "square matrix" => 8-element BenchmarkTools.BenchmarkGroup:
	  tags: []
	  "10 x 10" => TrialJudgement(-5.48% => improvement)
	  "100 x 100" => TrialJudgement(-30.11% => improvement)
	  "50 x 50" => TrialJudgement(-16.16% => improvement)
	  "200 x 200" => TrialJudgement(-45.62% => improvement)
	  "1000 x 1000" => TrialJudgement(-73.54% => improvement)
	  "400 x 400" => TrialJudgement(-58.71% => improvement)
	  "800 x 800" => TrialJudgement(-71.12% => improvement)
	  "2000 x 2000" => TrialJudgement(-74.34% => improvement)
```

### CODE REVIEW

**Comment by Gnimuc** [comment]
```julia
2-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Munkres" => 8-element BenchmarkTools.BenchmarkGroup:
	  tags: []
	  "10 x 10" => Trial(102.671 μs)
	  "100 x 100" => Trial(3.921 ms)
	  "50 x 50" => Trial(977.296 μs)
	  "200 x 200" => Trial(20.839 ms)
	  "1000 x 1000" => Trial(4.050 s)
	  "400 x 400" => Trial(194.381 ms)
	  "800 x 800" => Trial(1.975 s)
	  "2000 x 2000" => Trial(39.004 s)
  "Hungarian" => 8-element BenchmarkTools.BenchmarkGroup:
	  tags: []
	  "10 x 10" => Trial(12.428 μs)
	  "100 x 100" => Trial(1.922 ms)
	  "50 x 50" => Trial(297.304 μs)
	  "200 x 200" => Trial(8.417 ms)
	  "1000 x 1000" => Trial(784.957 ms)
	  "400 x 400" => Trial(54.454 ms)
	  "800 x 800" => Trial(445.907 ms)
	  "2000 x 2000" => Trial(6.305 s)
```
It looks like we can finish a `2000x2000` assignment problem in about `6s` comparing to current master's `25s`. The only bottleneck is finding the minimum:

https://github.com/Gnimuc/Hungarian.jl/blob/a53e23c9851a7540c7221163ec14a1ab2d13fec4/src/Munkres.jl#L297-L306

I feel like it still can be further improved(e.g. use multi-threading), but I've not test it yet. 

cc @dourouc05 Any feedback on this?

**Comment by dourouc05** [comment]
Indeed, this new version works very well! On my use cases (comparing running times of my whole algorithms, not just the Hungarian part), I gain 15-20% on my lightest uses, and even 25-30% on the heaviest! (With 100x100 problems.)

In my own benchmarks, I saw that adding elements to `Zs` (https://github.com/Gnimuc/Hungarian.jl/blob/master/src/Munkres.jl#L56) took quite a lot of time in reallocating (I really focused on memory, as my Julia processes tend to get killed for using too much RAM…). However, I could not find to preallocate a given number of items (with something like `sizehint!`). 

By the way, would you have an idea of a way to perform warm-starting for this problem? (Starting from a solution close to the optimum and performing a few iterations to reach optimality.) I know I could use a LP solver for this, but I'd be interested in a solution within the Hungarian algorithm. Not sure there is something easy to do, though :/.

**Comment by Gnimuc** [comment]
I don't think the reallocating of `Zs` here is the bottleneck. Since in most case `Zs` is very sparse, once it gets a large enough buffer, no extra reallocation occurs. The following benchmarks show that [the `find` functions](https://gist.github.com/Gnimuc/7ee2c27755b0bb410f9e82b58dab6c82#file-munkres-jl-mem-L292) is the one to blame. I guess those indices could be tracked in fixed-size vectors so as to get rid of the huge reallocation.
```julia
julia> using Hungarian, BenchmarkTools
julia> A = rand(100, 100);
julia> @benchmark hungarian($A)
BenchmarkTools.Trial: 
  memory estimate:  822.06 KiB
  allocs estimate:  2690
  --------------
  minimum time:     1.803 ms (0.00% GC)
  median time:      1.834 ms (0.00% GC)
  mean time:        1.918 ms (2.36% GC)
  maximum time:     4.059 ms (30.67% GC)
  --------------
  samples:          2604
  evals/sample:     1
```
Indeed, it's a bit wasteful to do a `100x100` assignment problem with about 1m memory, which leads to 30% gc time in the worst case.

```julia
# julia --track-allocation=user
using Hungarian
A = rand(100, 100);
for i = 1:100
    hungarian(A)
end
Profile.clear_malloc_data()
for i = 1:100
    hungarian(A)
end
exit()
```
generated reports: https://gist.github.com/Gnimuc/7ee2c27755b0bb410f9e82b58dab6c82

> By the way, would you have an idea of a way to perform warm-starting for this problem? (Starting from a solution close to the optimum and performing a few iterations to reach optimality.) I know I could use a LP solver for this, but I'd be interested in a solution within the Hungarian algorithm. Not sure there is something easy to do, though :/.

Unfortunately, I've not dived into that deep yet. This algorithm dates back to the 1950s(about 60 years ago :scream: :scream: :scream:), it's old but classical. I guess if there is a way to perform warm-starting, some variation of the algorithm should have existed and been published in a paper. The motivation of this package is to do exactly what J. Munkres told in his paper and make the implementation as fast as possible. To me, finding a good initialization is usually used in some convex optimization or gradient descent algorithms where it's a bit easier to get the intuition and heuristic/prior knowledge for warm-starting(well, maybe [not that much easier](https://en.wikipedia.org/wiki/Fast_inverse_square_root)😉). In the world of combinational optimization, finding such good initialization often causes a headache. :dizzy_face:

**Comment by Gnimuc** [comment]
With the latest commit 5b88468, a `100x100` assignment problem costs about `170k` memory comparing to `822k` before(about 4x smaller). A detailed report can be found [here](https://gist.github.com/Gnimuc/ac875c62a08ec5a925178a9c04b89a9d).
```julia
julia> using Hungarian, BenchmarkTools

julia> A = rand(100, 100);

julia> @benchmark hungarian($A)
BenchmarkTools.Trial: 
  memory estimate:  170.84 KiB
  allocs estimate:  575
  --------------
  minimum time:     1.641 ms (0.00% GC)
  median time:      1.725 ms (0.00% GC)
  mean time:        1.860 ms (0.84% GC)
  maximum time:     5.703 ms (47.48% GC)
  --------------
  samples:          2684
  evals/sample:     1
```

This also leads to another 10%~20% performance improvement:
```julia
File results of this run? (commit=a53e23, resultsdir=/Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results) (Y/n) y
INFO: Results of the benchmark were written to /Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results/a53e23c9851a7540c7221163ec14a1ab2d13fec4.jld
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "square matrix" => 8-element BenchmarkGroup([])

julia> showall(ans)
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "square matrix" => 8-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "10 x 10" => TrialJudgement(-10.16% => improvement)
          "100 x 100" => TrialJudgement(-11.94% => improvement)
          "50 x 50" => TrialJudgement(-9.87% => improvement)
          "200 x 200" => TrialJudgement(-14.55% => improvement)
          "1000 x 1000" => TrialJudgement(-17.99% => improvement)
          "400 x 400" => TrialJudgement(-17.31% => improvement)
          "800 x 800" => TrialJudgement(-21.34% => improvement)
          "2000 x 2000" => TrialJudgement(-10.39% => improvement)
```

**Comment by Gnimuc** [comment]
Now, the `munkres!` on this branch only takes `22K` memory :wink:,  @dourouc05  does this meet your use case?
 
```julia
julia> using Hungarian, BenchmarkTools
julia> A = rand(100, 100);
julia> @benchmark hungarian($A)
BenchmarkTools.Trial: 
  memory estimate:  108.00 KiB
  allocs estimate:  91
  --------------
  minimum time:     1.020 ms (0.00% GC)
  median time:      1.083 ms (0.00% GC)
  mean time:        1.161 ms (0.79% GC)
  maximum time:     4.898 ms (61.11% GC)
  --------------
  samples:          4296
  evals/sample:     1

julia> @benchmark Hungarian.munkres!($A)
BenchmarkTools.Trial: 
  memory estimate:  21.91 KiB
  allocs estimate:  64
  --------------
  minimum time:     1.009 ms (0.00% GC)
  median time:      1.032 ms (0.00% GC)
  mean time:        1.141 ms (0.16% GC)
  maximum time:     5.275 ms (55.10% GC)
  --------------
  samples:          4376
  evals/sample:     1
```


```julia
julia> using PkgBenchmark
julia> judge("Hungarian", "speed-up", "master")
INFO: Resolving dependencies for benchmark
INFO: No packages to install, update or remove
INFO: Running benchmarks...
Using benchmark tuning data in /Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/.tune.jld
INFO: No packages to install, update or remove
File results of this run? (commit=f0abfc, resultsdir=/Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results) (Y/n) y
INFO: Results of the benchmark were written to /Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results/f0abfc47090d5420f87bf51955412dcafbf83508.jld
INFO: Resolving dependencies for benchmark
INFO: No packages to install, update or remove
INFO: Running benchmarks...
Using benchmark tuning data in /Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/.tune.jld
INFO: No packages to install, update or remove
File results of this run? (commit=04d0fe, resultsdir=/Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results) (Y/n) y
INFO: Results of the benchmark were written to /Users/gnimuc/.julia/v0.6/.benchmarks/Hungarian/results/04d0fed0035a2bcce9923fbefd3133a9de8e3f2b.jld
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "square matrix" => 8-element BenchmarkGroup([])

julia> showall(ans)
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "square matrix" => 8-element BenchmarkTools.BenchmarkGroup:
          tags: []
          "10 x 10" => TrialJudgement(-41.13% => improvement)
          "100 x 100" => TrialJudgement(-52.68% => improvement)
          "50 x 50" => TrialJudgement(-54.51% => improvement)
          "200 x 200" => TrialJudgement(-63.45% => improvement)
          "1000 x 1000" => TrialJudgement(-81.61% => improvement)
          "400 x 400" => TrialJudgement(-73.59% => improvement)
          "800 x 800" => TrialJudgement(-80.13% => improvement)
          "2000 x 2000" => TrialJudgement(-81.01% => improvement)
```

File: src/Munkres.jl [hunk 1]
```diff
@@ -26,61 +26,77 @@ julia> full(matching)
...
 """
-    munkres(costMat) -> Zs
+    munkres!(costMat) -> Zs
 
 Identical to `munkres`, except that it directly modifies its input matrix `costMat`
-instead of allocating a copy. *As a result, the value of this matrix in the caller 
-code will be modified and should therefore no more be used!* This function should 
-rather be used by advanced users to improve performance of critical code. 
+instead of allocating a copy. As a result, the value of this matrix in the caller
+code will be modified and should therefore no more be used!* This function should
```
Comment (dourouc05) [inline src/Munkres.jl]: There is still an asterisk left here.

**Comment by dourouc05** [comment]
Those seem really nice improvements! I see a more than 50% reduction of memory allocations with my previous tests. Overall, this branch makes some tests go 50% faster than the current release! 

(Warm-starting with a LP solver still makes my code go faster for certain cases, up to 50% with respect to this PR, but I have seen no reference to it in the literature for the Hungarian algorithm, and it seems I am not alone ;). Unfortunately, Knuth has not written this part of his Art of Computer Programming yet!)

**Comment by dourouc05** [comment]
Great, thanks!

**Comment by Gnimuc** [comment]
Can't believe I didn't know that book! Time to buy one.

**Pull request merged**